### ISSUE TITLE
Reproducing Downstream Zero-Shot Performance of FairSeq-MoE models

### ISSUE THREAD

**Issue by rohitdwivedula** [facebookresearch/fairseq#4997]
## ❓ Questions and Help

In [this Arxiv paper](https://arxiv.org/pdf/2112.10684.pdf) describing FairSeq's MoE models, Table 2 (page 7) summarises the zero-shot performance of MoE models on various datasets. Is the script used to perform this analysis available in this repo? 

#### What I have done so far
For simplicity, I will describe what I have attempted to do so far with one of the datasets mentioned in the paper - PIQA. After downloading the PIQA dataset JSON file, I tokenized it into the format expected by Fairseq through the `MMapIndexedDatasetBuilder`.  I stored the binarized dataset in the folder `$MOUNT_PATH/downstream_data/FAIRSEQ/`. PIQA is a sentence completion task - so given a `prefix` and two possible suffixes `sol1` and `sol2`, you need to predict the correct completion. So, for each item in the dataset, I added two items to the Fairseq binarized dataset:
- `prefix || sol1`
- `prefix || sol2`
where `||` is the concatenation operator. Now, I would want to use the MoE model I trained to calculate the probabilities of each of these sentences and then use the word probabilities of suffixes `sol1` and `sol2` to determine which suffix to choose. (choose suffix which is more probable).

Now, I attempted to run a command that looks like this: 
```bash
NUM_GPUS=`ls /dev/ | grep -E 'nvidia[0-9]+' | wc -l`
CHECKPOINT_TO_PROCESS='checkpoint_last'
TOKENS_PER_SAMPLE=1024
BATCH_SIZE=1
MODEL_CAPACITY=32 
MOE_EVAL_CAPACITY_TOKEN_FRACTION=`python3 -c "print($MODEL_CAPACITY/($BATCH_SIZE * $TOKENS_PER_SAMPLE))"` 
DATA_PATH=$MOUNT_PATH/downstream_data/FAIRSEQ/

set -ux
python -m fairseq_cli.eval_lm \
  $DATA_PATH \
  --ddp-backend fully_sharded \
  --path $CHECKPOINT_FOLDER/$CHECKPOINT_TO_PROCESS.pt \
  --gen-subset PIQA \
  --tokens-per-sample $TOKENS_PER_SAMPLE
  --sample-break-mode none \
  --batch-size $BATCH_SIZE \
  --fp16 \
  --is-moe \
  --distributed-world-size $NUM_GPUS \
  --model-overrides "{'world_size': $NUM_GPUS, 'moe_eval_capacity_token_fraction': $MOE_EVAL_CAPACITY_TOKEN_FRACTION}" \
  --log-format json --output-word-probs
```

The last few lines of output of this command looks something like this: 
```bash
2023-02-28 09:31:30 | INFO | fairseq.moe_checkpoint_utils | Found total 64 expert files and current distributed world size: 8, Stitching experts to able to load on current world size.
.
.
.
2023-02-28 09:42:18 | INFO | fairseq_cli.eval_lm | 116 . [-2.865278]	</s> [-3.063847]	how [-8.623393]	Ġdo [-2.291624]	Ġyou [-0.984171]	Ġmake [-2.715627]	Ġa [-1.054973] ... (there are 1024 more tokens with their probabilities in this line)
2023-02-28 09:42:18 | INFO | fairseq_cli.eval_lm | eval_lm.py: Starting batch 1.
2023-02-28 09:42:18 | INFO | fairseq_cli.eval_lm | 86 ura [-6.828831]	ĠBit [-5.352273]	ters [-0.014752]	Ġ [-7.901725]	Ġ [-2.143599]	Ġ .... 
2023-02-28 09:42:18 | INFO | fairseq_cli.eval_lm | eval_lm.py: Starting batch 2.
.
.
.
2023-02-28 09:42:24 | INFO | fairseq_cli.eval_lm | Evaluated 122,578 tokens in 136.9s (895.31 tokens/s)
2023-02-28 09:42:24 | INFO | fairseq_cli.eval_lm | PIQA_basic Loss (base 2): 3.6791, Perplexity: 12.81
```

**Problems**: 
1. The log-probabilities of all tokens are **not** being outputted. The PIQA dataset has 119,506 tokens it, which I calculated by doing:
```python3
from fairseq.data import MMapIndexedDataset
ds = MMapIndexedDataset("./PIQA")
sum(map(lambda x: x.shape[0], ds)) # value is 119506
```
2. However, only log-probabilities for 15,058 tokens are being printed in the output. I am guessing that this is some issue of only one rank printing probabilities or something like that since 15,058 * 8 = 120,464, approximately equal to the total tokens in the dataset. 
3. Additionally, it looks like the data points are (a) being shuffled and (b) some data-points in the downstream dataset are being spliced across batches, which is a problem since the next batch starts in the middle of a sentence. Is there any way to prevent (a) and (b) from happening? 

If the scripts used to evaluate zero-shot performance on these datasets is available, would it be possible to have access to that? (the problems I listed above could probably be fixed by going through `eval_lm.py`, its imports and making changes, but want to avoid reinventing the wheel if possible) 

#### What's your environment?
 - fairseq version: moe branch
 - PyTorch Version: 1.8.0
 - OS: Linux
 - How you installed fairseq: From source
 - Build command you used: `cd fairseq && git checkout moe && pip install --upgrade pip && pip install --editable .`
 - Python version: 3.8.8
 - CUDA/cuDNN version:  CUDA Version: 11.1, CuDNN version: 8.0.5 [1]
 - GPU models and configuration: Nvidia Tesla V100 32GB * 8
# References
[1] CuDNN version found by looking at `cudnn_version.h`: 
```cpp
// cudnn_version.h
#define CUDNN_MAJOR 8
#define CUDNN_MINOR 0
#define CUDNN_PATCHLEVEL 5
```
